bolt-simple-taproot.md

```diff
@@ -0,0 +1,958 @@
+```
+
+The first segment uses a spare bit of the `leaf_version` to encode the parity
+of the output key, which is checked during control block verification. The
+`internal_key` is just that. The `inclusion_proof` is the series of sibling
+hashes in the path from the revealed leaf all the way up to the root of the
+tree. In practice if one hashes the revealed leaf, and each 32-byte hash of the
+inclusion proof together, they'll reach the script root if the proof was valid.
+
+If only a single leaf is committed to in a tree, then the `inclusion_proof`
+will be absent.
+
+The final witness to spend a script path output is:
+```
+<witness1> ... <witnessN> <leaf_script> <control_block>
+```
+
+### MuSig2
+
+Musig2 is a multi-signature scheme based on schnorr signatures. It allows N
+parties to construct an aggregated public key, and then generate a single
+multi-signature that can only be generated with all the parties of the
+aggregated key. In practice we use this to allow the funding output of
+channel to look just like any other P2TR output.
+
+A musig2 signing flow has 4 stages:
+
+  1. First each party generates a public nonce, and exchanges it with every
+    other party.
+
+  2. Next public keys are exchanged by each party, and combined into a single
+    aggregated public key.
+
+  3. Next, each party generates a _partial_ signature and exchanges it with
+    every other party.
+
+  4. Finally, once all partial signatures are known, they can be combined into a
+    valid schnorr signture, which can be veirfied using BIP 340.
+
+Steps 1&2 may be performed out of order, or concurrently. In our case only two
+parties exist, so as soon as one party knows both partial signatures, they can
+be combined into a final signature.
+
+#### Key Aggregation
+
+We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)
+for key aggregation. In order to avoid sending extra ordering information, we
+always assume that both keys are sorted first with the `KeySort` algorithm
+before aggregating (`KeyAgg(KeySort(p1, p2))`)>
+
+#### Nonce Generation
+
+A `musig2` public nonce is a 66-byte value which is the concentration of two EC
+points encoded using the compressed format:
+```
+nonce_1 || nonce_2
+```
+
+Each public nonce has a corresponding _secret_ nonce which is only known to the
+signer.
+
+For nonce generation, we refer to the `NonceGen` algorithm defined in the
+`musig2` BIP. We only specify the set of required arguments, though a signer
+can also opt to specify the additional argument in order to strengthen their
+nonce.
+
+#### Nonce Handling
+
+Throughput this document we always assume that _fresh_ nonces are generated for
+each signing attempt, and nonces are **never** written to disk. As we'll see
+later, this has implications with respect to the retransmission protocol.
+
+Note that nonces can safely be pre-generated, and even batched/pipelined ahead
```

Comment:
It might be worth noting that to "safely" pre-generate the nonces, some care needs to be taken to ensure a pre-generated nonce is never reused. As noted in the musig [doc](https://github.com/ElementsProject/secp256k1-zkp/blob/master/src/modules/musig/musig.md): Pre-generating nonces "disables some of the defense-in-depth measures that may protect against API misuse in some cases."

Reply:
a straight link to the BIP is probably a good idea here to know exactly what should be raising red flags when implementing. e.g. I find it highly sketchy if someone thinks they need to serialize the secret nonce.

Reply:
Agreed, it ultimately depends on what the `musig` API of the implementations looks like. By "pre-generate" nonces, I mean generating your nonce before you know the public key of the remote party. 

Will update this with some links to the BIP, and maybe also will start to directly reference some of the algos there.

Reply:
Tacked on a section here describing the purely random nonce generation, and also a way to used the existing commitment height/number and the shachain scheme to generate a unique nonce for each channel state.